model-sync-gradle/src/main/kotlin/org/modelix/model/sync/gradle/ModelSyncGradlePlugin.kt

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+    }
+
+    private fun registerTasksForSyncDirection(
+        syncDirection: SyncDirection,
+        project: Project,
+        previousTask: TaskProvider<*>,
+    ) {
+        val baseDir = project.buildDir.resolve("model-sync").apply { mkdirs() }
+        val jsonDir = baseDir.resolve(syncDirection.name).apply { mkdir() }
+        val sourceTask = when (syncDirection.source) {
+            is LocalSource -> registerTasksForLocalSource(syncDirection, project, previousTask, jsonDir)
+            is ServerSource -> registerTasksForServerSource(syncDirection, project, previousTask, jsonDir)
+            else -> previousTask
+        }
+
+        when (syncDirection.target) {
+            is LocalTarget -> registerTasksForLocalTarget(syncDirection, project, sourceTask, jsonDir)
+            is ServerTarget -> registerTasksForServerTarget(syncDirection, project, sourceTask, jsonDir)
+        }
+    }
+
+    private fun registerTasksForServerSource(
+        syncDirection: SyncDirection,
+        project: Project,
+        previousTask: TaskProvider<*>,
+        jsonDir: File,
+    ): TaskProvider<*> {
+        val serverSource = syncDirection.source as ServerSource
+
+        val name = "${syncDirection.name}ExportFromModelServer"
+        val exportFromModelServer = project.tasks.register(name, ExportFromModelServer::class.java) {
+            it.dependsOn(previousTask)
+            it.outputDir.set(jsonDir)
+            it.url.set(serverSource.url)
+            it.repositoryId.set(serverSource.repositoryId)
+            it.branchName.set(serverSource.branchName)
+            it.revision.set(serverSource.revision)
+        }
+        return exportFromModelServer
+    }
+
+    private fun registerTasksForLocalSource(
+        syncDirection: SyncDirection,
+        project: Project,
+        previousTask: TaskProvider<*>,
+        jsonDir: File,
+    ): TaskProvider<*> {
+        val localSource = syncDirection.source as LocalSource
+        val antScript = jsonDir.resolve("build.xml")
+        val generateAntScript = project.tasks.register(
+            "${syncDirection.name}GenerateAntScriptForExport",
+            GenerateAntScriptForMps::class.java,
+        ) {
+            it.dependsOn(previousTask)
+            it.mpsHomePath.set(localSource.mpsHome?.absolutePath)
+            it.mpsHeapSize.set(localSource.mpsHeapSize)
+            it.repositoryPath.set(localSource.repositoryDir?.absolutePath)
+            it.antScriptFile.set(antScript)
+            it.mpsDependenciesPath.set(getDependenciesDir(project).absolutePath)
+            it.jsonDirPath.set(jsonDir.absolutePath)
+            it.exportFlag.set(true)
+            it.includedModules.set(syncDirection.includedModules)
+        }
+
+        val exportFromMps = project.tasks.register("${syncDirection.name}ExportFromMps", ExportFromMps::class.java) {
+            it.dependsOn(generateAntScript)
+            it.outputs.cacheIf { false }
```

Comment:
Can this be optimized later on? In theory we could hash the full source MPS project and use that as a cache key?